aiida/engine/daemon/execmanager.py

```diff
@@ -199,13 +201,26 @@ def upload_calculation(
         if data_node is None:
             logger.warning(f'failed to load Node<{uuid}> specified in the `local_copy_list`')
         else:
-            dirname = os.path.dirname(target)
-            if dirname:
-                os.makedirs(os.path.join(folder.abspath, dirname), exist_ok=True)
-            with folder.open(target, 'wb') as handle:
-                with data_node.open(filename, 'rb') as source:
-                    shutil.copyfileobj(source, handle)
-            provenance_exclude_list.append(target)
+
+            # If no explicit source filename is defined, we assume the top-level directory
+            filename_source = filename or '.'
+            filename_target = target or ''
+
+            # Make the target filepath absolute and create any intermediate directories if they don't yet exist
+            filepath_target = pathlib.Path(folder.abspath) / filename_target
+            filepath_target.parent.mkdir(parents=True, exist_ok=True)
+
+            if data_node.get_object(filename_source).file_type == FileType.DIRECTORY:
+                # If the source object is a directory, we copy its entire contents
+                data_node.copy_tree(filepath_target, filename_source)
+                provenance_exclude_list.extend(data_node.list_object_names(filename_source))
+            else:
+                # Otherwise, simply copy the file
+                with folder.open(target, 'wb') as handle:
+                    with data_node.open(filename, 'rb') as source:
+                        shutil.copyfileobj(source, handle)
+
+                provenance_exclude_list.append(target)
```

Comment:
Do we need this conditional? I think `data_node.copy_tree` should work when it is provided with the path to a single file, no? (actually it might be good to add a test for this specifically in #5114 )

Same question for `data_node.list_object_names`; even if it currently doesn't, maybe it would be easier to make that work with paths to files instead of adding complexity here.

Reply:
Because I don't `copy_tree` should accept a file for a `path`. I will add an explicit check in the other PR and add a test.

Reply:
Mmm ok, I see. I guess I understand how "listing" is technically different if interpreted as "show the content of" and that doesn't work with files, but "copy" is more general and works with both. Is it the "tree" qualifier that for you implies that what is copied should be a directory and files should be identified and rejected? Because I see no practical complications for copying files and folders indistinctively...

Also, related to this: what would be the problem of doing `provenance_exclude_list.append(target)` when `target` is a folder? I mean, when you do `data_node.list_object_names(filename_source))`, the list you get back also includes subfolders, right? So you are already adding folders to the `exclude_list`.

On the other hand, I guess one problem is that if you add a subfolder like that and then create some input file that should be included in the provenance inside of that subfolder, then it will be ignored because the folder is on the exclude list, right? But we are anyways adding subfolders because of previous paragraph, so this problem is more complex and not solved by this `if`.

Reply:
> Mmm ok, I see. I guess I understand how "listing" is technically different if interpreted as "show the content of" and that doesn't work with files, but "copy" is more general and works with both. Is it the "tree" qualifier that for you implies that what is copied should be a directory and files should be identified and rejected? Because I see no practical complications for copying files and folders indistinctively...

With the name `copy_tree`, I think it should definitely not accept files. The `tree` indeed signifies that it operates on a directory, like `put_object_from_tree`. For files there is `put_object_from_file(like)`. It can be argued that there should be a generic method that copies both, but I don't think it necessarily makes things easier to use or implement.

> Also, related to this: what would be the problem of doing `provenance_exclude_list.append(target)` when `target` is a folder? I mean, when you do `data_node.list_object_names(filename_source))`, the list you get back also includes subfolders, right? So you are already adding folders to the `exclude_list`.

You are correct. Since I expand the filenames in their separate subpaths when they are copied over to the node's repository, this will also include the subpath that equals `target`. So there is indeed no need to expand the target path already here in its subfolders. I will change this.

> On the other hand, I guess one problem is that if you add a subfolder like that and then create some input file that should be included in the provenance inside of that subfolder, then it will be ignored because the folder is on the exclude list, right? But we are anyways adding subfolders because of previous paragraph, so this problem is more complex and not solved by this `if`.

This is a good point. I guess this might indeed be problematic, but I don't see a clean solution right now. If the file really needs to be stored in the node's repo, then they should simply write the entire folder directly to the sandbox and not use the `local_copy_list`. Anyway, this entire system of the `local_copy_list` is not an ideal solution but it is there for historical reasons. Really what we should have done from the beginning is just have the plugin write everything to the sandbox and then specify just the `provenance_from_exclude` list. This gives full flexibility in terms of what gets stored and what not and it is a lot simpler to understand. This is definitely a thing we should consider simplifying for v3.0 if that ever comes out.

Reply:
Actually, thinking about it, we cannot simplify this by just adding `target` to the `provenance_exclude_list`. The reason is that if the target is `.`, i.e., add all contents to the root, we don't want to exclude everything that is in there. Just those resources that were written by that operation. So what I had originally, expanding the source path into its contents, is the correct thing to do.

Reply:
Uhm, I see. I still feel a bit uneasy about this, but I think this is a general problem of implementing file copy interfaces. It kind of reminds me of when I tried to add directory creation to the copy lists and we had to roll it back because it changed the behavior of how some copy commands were indicated, prompting the publication of issue #4566 out of my frustration. What a headache =/...

Anyways, I still think it would be good to do a more thorough design of this and define a specific interface, but that is beyond the scope of this PR, so I'm just going to approve this now.

Reply:
I agree that the interface is shit, but the real solution is to get rid of `local_copy_list` entirely. The `remote_copy_list` is slightly different, but ideally that could also be replaced with an interface that allows a user to _directly_ manipulate the remote file system instead of giving abstract indirect instructions.

Reply:
I don't know, I get the feeling that you will get this kind of annoying headaches with almost any sort of general "file copying" interface you end up implementing. I mean, just from writing bash scripts with `cp` or `rsync` I remember having problems with suddenly unexpected behavior related to the exact syntax used in the command and the specifics of the situation. I guess that is the reason why you or shutil choose to do `copy_TREE` kind of methods that are more restrictive (instead of what I wanted to just copy whatever you provide) but at least the expected result is more clear.

> The `remote_copy_list` is slightly different, but ideally that could also be replaced with an interface that allows a user to _directly_ manipulate the remote file system instead of giving abstract indirect instructions.

I don't understand, what do you mean by _directly_ manipulating the remote file system?

Reply:
> I don't understand, what do you mean by directly manipulating the remote file system?

By letting the user call direct methods to write/read files instead of using indirect instructions through `remote_copy_list`. It would automatically get a transport, which can then be used to perform the operations directly. This would ideally have an interface that is identical to writing files to the local file system, making it transparent whether you are reading/writing from/to a local or remote file system.